src/table-index.ts

```diff
@@ -0,0 +1,148 @@
+import { Column, ColumnDefinition, ColumnDefinitionsToColumns } from './column';
+import { DefaultExpression, Expression } from './expression';
+import { toSnakeCase, wrapQuotes } from './naming';
+import { GroupToken, SeparatorToken, StringToken, Token } from './tokens';
+
+export type IndexDefinitionsToIndexes<
+  TableNameT extends string,
+  IndexDefinitionsT extends { [column: string]: IndexDefinition<boolean, boolean> },
+> = {
+  [IndexName in keyof IndexDefinitionsT]: IndexName extends string
+    ? IndexDefinitionsT[IndexName] extends IndexDefinition<infer IsPrimaryKey, infer IsUniqueKey>
+      ? Index<IndexName, TableNameT, IsPrimaryKey, IsUniqueKey>
+      : never
+    : never;
+};
+
+export interface IndexDefinitionFormat {
+  type: string;
+  expressions: Expression<any, boolean, string>[];
+  isUniqueKey: boolean;
+  isPrimaryKey: boolean;
+  include: Column<string, string, any, boolean, boolean, any>[];
+  where: DefaultExpression<boolean> | null;
+}
+
+export interface IndexDefinition<
+  IsPrimaryKey extends boolean = false,
+  IsUniqueKey extends boolean = false,
+> {
+  getDefinition(): IndexDefinitionFormat;
+  where(condition: DefaultExpression<boolean>): IndexDefinition<IsPrimaryKey, IsUniqueKey>;
+  primaryKey(): IndexDefinition<true, true>;
+  unique(): IndexDefinition<IsPrimaryKey, true>;
+  include(
+    ...columns: Column<string, string, any, boolean, boolean, any>[]
+  ): IndexDefinition<IsPrimaryKey, IsUniqueKey>;
+}
+
+export class Index<
+  Name extends string,
```

Comment:
Why is it useful to have these as type parameters? (Not having them as type parameters would make some things simpler.)

Reply:
Mostly answered in the other comment, but basically they're not that useful and we can drop them.